### ISSUE TITLE
Key lag using "event" macros

### ISSUE THREAD

**Issue by StarterBA** [ChapelR/custom-macros-for-sugarcube-2#70]
Hello, ChapelR!

I have Twine 2.8.1, Sugarcube 2.36.1.

I encountered a lag/performance issue with the "event" macros. I tried to use it to reduce the clicking on links. So I added the keybinds "I" for inventory, "C" for characteristics, "R" for return links into StoryMenu passage. But when I press the keys, the number of turns increases by tens or sometimes 20-30, so there is lag. The more I use keys, the more are turns and lagging. When I click with mouse everything is ok, 1 turn per click, no lag.

Maybe I added the code to wrong place? 
I thought that maybe I should try adding a widget to each passage instead.

Thanks in advance!
[TwineIssue.zip](https://github.com/user-attachments/files/16415779/TwineIssue.zip)

**Comment by ChapelR** [comment]
There is one major and one minor issue with your code here. The major issue that is causing your problems is that you placed the `<<on>>` macro in `StoryMenu`, which is a passage that is reprocessed every turn, meaning that the number of event listeners increases by 1 for each key press for every turn that happens. So basically, by the time you move through even just 3 or 4 passages, a single key press is triggering 3 or 4 passage transitions in a row. I strongly recommend moving your `<<on>>` macros from `StoryMenu` to `StoryInit`.

The more minor issue is the use of the `"keydown"` event type. You are better off using `"keyup"` for a few reasons. For one, `"keydown"` can be triggered repeatedly by holding down a key on the keyboard, which can cause problems. Second, `"keyup"` reflects the way most keypresses in most other apps are detected, so will match the expectations of most users more closely. 

TL;DR:

- **Must**: Move your `<<on>>` macros from `StoryMenu` to `StoryInit`.  
- **Recommendation**: Switch the event type from `"keydown"` to `"keyup"`.

**Comment by greyelf** [comment]
@ChapelR 
You beat my reply by one minute! lol

**Comment by StarterBA** [comment]
Thanks a LOT!
Moved the code to StoryInit. Now everything works as planned!
